File: tpx/private/tpx/I3VEMConverter.cxx
/*
 * copyright  (C) 2011
 * The Icecube Collaboration
 *
 * $Id$
 *
 * @version $Revision$
 * @date $LastChangedDate$
 * @author Fabian Kislat <fabian.kislat@desy.de> Last changed by: $LastChangedBy$
 */

#include "utility.h"

#include <icetray/I3ConditionalModule.h>
#include <icetray/I3Frame.h>
#include <icetray/OMKey.h>
#include <icetray/I3Units.h>
#include <dataclasses/calibration/I3Calibration.h>
#include <dataclasses/physics/I3RecoPulse.h>

#include <boost/foreach.hpp>
#include <string>


class I3VEMConverter : public I3ConditionalModule {
public:
  I3VEMConverter(const I3Context &ctx)
    : I3ConditionalModule(ctx)
  {
    pePulsesName_ = "";
    AddParameter("PEPulses", "Input pulses with charge in PE units", pePulsesName_);
    vemPulsesName_ = "";
    AddParameter("VEMPulses", "Output pulses, will be calibrated in VEM", vemPulsesName_);
  }

  void Configure();

  void DAQ(I3FramePtr frame);

private:
  std::string pePulsesName_;
  std::string vemPulsesName_;

  SET_LOGGER("I3VEMConverter");
};

I3_MODULE(I3VEMConverter);


void I3VEMConverter::Configure()
{
  GetParameter("PEPulses", pePulsesName_);
  if (pePulsesName_.empty()) {
    log_fatal("No input pulses specified.");
  }

  GetParameter("VEMPulses", vemPulsesName_);
  if (vemPulsesName_.empty()) {
    log_fatal("No name for the output pulses specified.");
  }

  if (vemPulsesName_ == pePulsesName_) {
    log_fatal("Output pulses have the same name as input pulses, "
	      "but must be different.");
  }
}


void I3VEMConverter::DAQ(I3FramePtr frame)
{
  I3RecoPulseSeriesMapConstPtr pePulses =
    frame->Get<I3RecoPulseSeriesMapConstPtr>(pePulsesName_);
  if (!pePulses) {
    log_info("Input pulses ('%s') not in frame. Skipping.",
	     pePulsesName_.c_str());
    PushFrame(frame);
    return;
  }

  const I3Calibration &calibration = frame->Get<I3Calibration>();

  // copy pulses
  I3RecoPulseSeriesMapPtr vemPulses(new I3RecoPulseSeriesMap(*pePulses));

  typedef std::pair<const OMKey, std::vector<I3RecoPulse> > PairOMKeyRecoPulseSeries;
  BOOST_FOREACH(PairOMKeyRecoPulseSeries &item, *vemPulses) {
    const OMKey &omKey = item.first;
    const I3VEMCalibration *vemcal = tpx::get_dom_info(omKey, calibration.vemCal);
    if (!vemcal) {
      log_error("Pulse in DOM %s, but no VEM calibration! Charge will be NAN.",
		omKey.str().c_str());
    }
    double pe_per_vem = vemcal ? vemcal->pePerVEM/vemcal->corrFactor : NAN;

    BOOST_FOREACH(I3RecoPulse &pulse, item.second) {
      pulse.SetCharge(pulse.GetCharge()/pe_per_vem);
    }
  }

  frame->Put(vemPulsesName_, vemPulses);
  PushFrame(frame);
}

File: icetop_Level3_scripts/python/modules/I3VEMConverter.py
from icecube import dataclasses
from icecube.icetray import I3ConditionalModule, I3Frame, I3Units
from icecube.dataclasses import I3WaveformSeriesMap


class I3VEMConverter(I3ConditionalModule):
    """I3Module to calibrate waveforms. This module replicates the
    calibration applied in tpx/I3TopHLCPulseExtractor, which does not
    add the VEM-calibrated waveform to the frame.
    """
    def __init__(self, ctx):
        I3ConditionalModule.__init__(self, ctx)
        self.AddParameter('Waveforms', 'Waveforms before VEM calibration', 'IceTopCalibratedWaveforms')
        self.AddParameter('VEMWaveforms', 'VEM-calibrated waveforms', 'IceTopVEMCalibratedWaveforms')
        self.AddOutBox("OutBox")

    def Configure(self):
        self.vem_waveforms_ = self.GetParameter('VEMWaveforms')
        self.pe_waveforms_ = self.GetParameter('Waveforms')

    def DAQ(self, frame):
        import numpy
        import copy
        if not 'I3Calibration' in frame:
            raise Exception("I3Calibration is not in the frame. I can't calibrate! Did you pass a GCD file?")
        if not 'I3DetectorStatus' in frame:
            raise Exception("I3DetectorStatus is not in the frame. I can't calibrate! Did you pass a GCD file?")

        status = frame['I3DetectorStatus']
        calibration = frame['I3Calibration']

        if not self.pe_waveforms_ in frame:
            self.PushFrame(frame)
            return

        waveforms = frame[self.pe_waveforms_]
        vem_waveforms = copy.deepcopy(waveforms)

        for omKey in waveforms.keys():
            if omKey in calibration.vem_cal and omKey in calibration.dom_cal and omKey in status.dom_status.keys():
                vemcal = calibration.vem_cal[omKey]
                domcal = calibration.dom_cal[omKey]
                dom_status = status.dom_status[omKey]
                pe_per_vem = vemcal.pe_per_vem/vemcal.corr_factor
                spe_mean = dataclasses.spe_mean(dom_status, domcal)
                scale = 1./spe_mean/domcal.front_end_impedance/pe_per_vem
            else:
                log_error('OM %s present in waveform keys but not in I3Calibration and I3DetectorStatus! Charge will be NAN'%omKey)
                scale = numpy.nan
            for launch in range(len(waveforms[omKey])):
                w_in = waveforms[omKey][launch].waveform
                w_out = vem_waveforms[omKey][launch].waveform
                for i in range(len(w_in)):
                    w_out[i] = w_in[i]*waveforms[omKey][launch].bin_width*scale

        frame.Put(self.vem_waveforms_, vem_waveforms)
        self.PushFrame(frame)
